File: .codex_review_extract/flutter-pos/lib/modules/shift/controllers/shift_controller.dart
import 'package:flutter_easyloading/flutter_easyloading.dart';
import 'package:get/get.dart';
import 'package:flutter/material.dart';
import '../../../app/services/api_client.dart';
import '../../../app/services/storage_service.dart';
import '../../../data/models/shift.dart';
import '../../../shared/widgets/app_dialog.dart';

/// UI CHANGE: `closeShift()` used to open a `Get.defaultDialog` — GetX's
/// own stock dialog helper, which renders as a plain white rounded card
/// with default text buttons. That's visually inconsistent with every
/// other confirmation in the app (delete confirmations, etc.), which all
/// go through [AppDialog.confirm] and get the glass panel + gradient CTA
/// treatment. Swapped to `AppDialog.confirm` for the same look and to
/// keep exactly one confirmation-dialog implementation in the codebase.
class ShiftController extends GetxController {
  final ApiClient _api = ApiClient.to;
  final activeShift = Rx<Shift?>(null);
  final pettyCashController = TextEditingController();
  final pettyController = TextEditingController();
  final loading = false.obs;

  @override
  void onInit() {
    super.onInit();
    loadActive();
  }

  @override
  void onClose() {
    pettyCashController.dispose();
    pettyController.dispose();
    super.onClose();
  }

  Future<void> loadActive() async {
    loading.value = true;
    final res = await _api.get('/shifts/active', fromData: (d) {
      return d == null ? null : Shift.fromJson(d as Map<String, dynamic>);
    });
    loading.value = false;
    if (res.success) {
      final data = res.data;
      activeShift.value = data;
      if (data != null) {
        await StorageService.to.saveShift(data.idShift);
      } else {
        await StorageService.to.clearShift();
      }
    }
  }

  Future<void> openShift() async {
    final petty = double.tryParse(pettyCashController.text) ?? 0;
    loading.value = true;
    EasyLoading.show(status: 'Opening...');
    final res = await _api.post('/shifts/open', body: {'petty_cash': petty});
    loading.value = false;
    EasyLoading.dismiss();
    if (res.success) {
      EasyLoading.showSuccess('Shift opened');
      await loadActive();
    } else {
      EasyLoading.showError(res.message);
    }
  }

  Future<void> addPettyCash() async {
    if (activeShift.value == null) return;
    final nominal = double.tryParse(pettyController.text) ?? 0;
    if (nominal <= 0) {
      EasyLoading.showError('Nominal required');
      return;
    }
    EasyLoading.show(status: 'Saving...');
    final res = await _api.post(
      '/shifts/${activeShift.value!.idShift}/petty-cash',
      body: {'nominal': nominal, 'keterangan': 'Petty cash'},
    );
    EasyLoading.dismiss();
    if (res.success) {
      pettyController.clear();
      EasyLoading.showSuccess('Petty cash recorded');
    } else {
      EasyLoading.showError(res.message);
    }
  }

  Future<void> closeShift() async {
    if (activeShift.value == null) return;
    final confirmed = await AppDialog.confirm(
      title: 'Close Shift',
      message: 'Are you sure you want to close this shift?',
      confirmText: 'Close',
    );
    if (confirmed != true) return;

    EasyLoading.show(status: 'Closing...');
    final res = await _api.post(
      '/shifts/${activeShift.value!.idShift}/close',
    );
    EasyLoading.dismiss();
    if (res.success) {
      EasyLoading.showSuccess('Shift closed');
      await loadActive();
    } else {
      EasyLoading.showError(res.message);
    }
  }
}

File: .codex_review_extract/laravel-api/app/Http/Controllers/Shift/ShiftController.php
<?php

namespace App\Http\Controllers\Shift;

use App\Http\Controllers\Controller;
use App\Http\Requests\Shift\OpenShiftRequest;
use App\Http\Requests\Shift\PettyCashRequest;
use App\Services\ShiftService;
use Illuminate\Http\JsonResponse;

class ShiftController extends Controller
{
    public function __construct(private ShiftService $service) {}

    public function active(): JsonResponse
    {
        $shift = $this->service->activeForUser(auth()->id());

        return $this->ok($shift, $shift ? 'Active shift found' : 'No active shift');
    }

    public function open(OpenShiftRequest $request): JsonResponse
    {
        try {
            $shift = $this->service->open(
                auth()->id(),
                (float) $request->input('petty_cash'),
                $request->ip()
            );
        } catch (\RuntimeException $e) {
            return $this->error($e->getMessage(), 422);
        }

        return $this->created($shift, 'Shift opened');
    }

    public function pettyCash(int $id, PettyCashRequest $request): JsonResponse
    {
        $petty = $this->service->addPettyCash(
            $id,
            (float) $request->input('nominal'),
            $request->input('keterangan')
        );

        return $this->created($petty, 'Petty cash recorded');
    }

    public function close(int $id): JsonResponse
    {
        $shift = \App\Models\Shift::findOrFail($id);

        try {
            $closing = $this->service->close($shift, request()->ip());
        } catch (\RuntimeException $e) {
            return $this->error($e->getMessage(), 422);
        }

        return $this->created($closing->load('shift'), 'Shift closed');
    }

    public function history(): JsonResponse
    {
        return $this->paginated($this->service->history((int) request()->query('perPage', 15)));
    }
}
